### PR TITLE
✨client: client.MatchingFields support multiple indexes

### PR DESCRIPTION
**Pull request by halfcrazy**
add support for multiple indexes when using client.MatchingFields

Fix https://github.com/kubernetes-sigs/controller-runtime/issues/612

### CODE REVIEW

**Comment by halfcrazy** [comment]
/hold 

After reading https://github.com/kubernetes-sigs/controller-runtime/pull/1838#pullrequestreview-908261250 seems we need an atomic operation here.

File: pkg/cache/internal/cache_reader.go [hunk 1]
```diff
@@ -178,6 +177,54 @@ func (c *CacheReader) List(_ context.Context, out client.ObjectList, opts ...cli
...
+			if len(objs) == 0 {
+				return nil, nil
+			}
+			continue
+		}
+		fn, exist := indexers[indexName]
+		if !exist {
+			return nil, fmt.Errorf("index with name %s does not exist", indexName)
+		}
+		tObjs := make([]interface{}, 0, len(objs))
```
Comment (alvaroaleman) [inline pkg/cache/internal/cache_reader.go:207]:
```suggestion
		filteredObjects := make([]interface{}, 0, len(objs))
```

Comment (halfcrazy) [reply pkg/cache/internal/cache_reader.go]: done

File: pkg/client/fake/client.go [hunk 2]
```diff
@@ -597,15 +595,24 @@ func (c *fakeClient) filterWithFields(list []runtime.Object, gvk schema.GroupVer
...
+			return nil, fmt.Errorf("List on GroupVersionKind %v specifies selector on field %s, but no "+
+				"index with name %s has been registered for GroupVersionKind %v", gvk, req.Field, req.Field, gvk)
+		}
 	}
 
-	indexExtractor := indexes[fieldKey]
 	filteredList := make([]runtime.Object, 0, len(list))
 	for _, obj := range list {
-		if c.objMatchesFieldSelector(obj, indexExtractor, fieldVal) {
+		ok := true
```
[the remark this answers is not in the document]
Comment (halfcrazy) [reply pkg/client/fake/client.go]: done

File: pkg/cache/cache_test.go [hunk 3]
```diff
@@ -1968,6 +1968,51 @@ func CacheTest(createCacheFunc func(config *rest.Config, opts cache.Options) (ca
...
+						client.MatchingFields{fieldName1: testNamespaceTwo})).To(Succeed())
+					Expect(listObj.Items).To(HaveLen(3))
+
+					By("listing pods with restart policy index")
+					listObj = &corev1.PodList{}
+					Expect(informer.List(context.Background(), listObj,
+						client.MatchingFields{fieldName2: string(corev1.RestartPolicyAlways)})).To(Succeed())
+					Expect(listObj.Items).To(HaveLen(2))
+
+					By("listing Namespaces with both fixed indexers 1 and 2")
```
[the remark this answers is not in the document]
Comment (halfcrazy) [reply pkg/cache/cache_test.go]: done

File: pkg/cache/cache_test.go [hunk 4]
```diff
@@ -1968,6 +1968,51 @@ func CacheTest(createCacheFunc func(config *rest.Config, opts cache.Options) (ca
...
+					listObj = &corev1.PodList{}
+					Expect(informer.List(context.Background(), listObj,
+						client.MatchingFields{fieldName1: testNamespaceTwo, fieldName2: string(corev1.RestartPolicyAlways)})).To(Succeed())
+					Expect(listObj.Items).To(HaveLen(2))
```
[the remark this answers is not in the document]
Comment (halfcrazy) [reply pkg/cache/cache_test.go]: Add a new pod7 with ns three to demonstrate filter works.

**Comment by halfcrazy** [comment]
The test pass in my local env.
```
# yan @ yan-virtualbox in ~/workspace/oss-code/controller-runtime on git:multiple-indexes o [11:25:09]
$ git rev-parse --short HEAD
ebaa704e

# yan @ yan-virtualbox in ~/workspace/oss-code/controller-runtime/pkg/cache on git:multiple-indexes o [11:12:52]
$ ginkgo run
Running Suite: Cache Suite - /home/yan/workspace/oss-code/controller-runtime/pkg/cache
======================================================================================
Random Seed: 1696821181

Will run 235 of 235 specs
•••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••

Ran 235 of 235 Specs in 76.783 seconds
SUCCESS! -- 235 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1m19.445324495s
Test Suite Passed
```

File: pkg/cache/internal/cache_reader_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package internal
```
Comment (alvaroaleman) [inline pkg/cache/internal/cache_reader_test.go:1]: Why the benchmark? This isn't suited if people care about performance, its just for convenience. If someome needs performance, they should be creating one index for all things they want to index, i.E. in your testexample a label+restart policy index

Comment (halfcrazy) [reply pkg/cache/internal/cache_reader_test.go:1]: Try to keep in mind there is a potential performance issue with multiple indexes. By tracking upstream `ByIndexes`, if upstream implement and we'll switch to standard `ByIndexes`.

Comment (alvaroaleman) [reply pkg/cache/internal/cache_reader_test.go:1]: Yes I see that - Which is why I think this benchmark is pointless. Anyone who cares about perf should not be using this.

Comment (halfcrazy) [reply pkg/cache/internal/cache_reader_test.go:1]: OK, I will delete it.

**Review by alvaroaleman — APPROVED** [review]
Thanks!

**Comment by tareksha** [comment]
hi @halfcrazy can you give a short explanation about the expected performance of this implementation? how does it grow in relation to: 1- number of objects in store, 2- number of selectors.

**Comment by halfcrazy** [comment]
> hi @halfcrazy can you give a short explanation about the expected performance of this implementation? how does it grow in relation to: 1- number of objects in store, 2- number of selectors. my local tests show somewhat linear growth by both..

In my local environment, there is an obvious step between 1 index and 2 indexes, and linear growth for more indexes.

```
cpu: Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
BenchmarkByIndexes1-4   	    1784	    763295 ns/op
BenchmarkByIndexes1-4   	    1692	    757617 ns/op
BenchmarkByIndexes1-4   	    1701	    617487 ns/op
BenchmarkByIndexes1-4   	    1837	    697619 ns/op
BenchmarkByIndexes1-4   	    1886	    655327 ns/op
BenchmarkByIndexes2-4   	     488	   2438093 ns/op
BenchmarkByIndexes2-4   	     440	   2744782 ns/op
BenchmarkByIndexes2-4   	     444	   2654877 ns/op
BenchmarkByIndexes2-4   	     451	   2733612 ns/op
BenchmarkByIndexes2-4   	     412	   2699114 ns/op
BenchmarkByIndexes3-4   	     373	   3046349 ns/op
BenchmarkByIndexes3-4   	     344	   3228925 ns/op
BenchmarkByIndexes3-4   	     319	   5556805 ns/op
BenchmarkByIndexes3-4   	      86	  15180030 ns/op
BenchmarkByIndexes3-4   	     289	   3911239 ns/op
BenchmarkByIndexes4-4   	     319	   3426624 ns/op
BenchmarkByIndexes4-4   	     348	   3536687 ns/op
BenchmarkByIndexes4-4   	     364	   3647029 ns/op
BenchmarkByIndexes4-4   	     289	   4686571 ns/op
BenchmarkByIndexes4-4   	     276	   4342592 ns/op
BenchmarkByIndexes5-4   	     309	   4256660 ns/op
BenchmarkByIndexes5-4   	     332	   3915095 ns/op
BenchmarkByIndexes5-4   	     307	   4555571 ns/op
BenchmarkByIndexes5-4   	     283	   3688207 ns/op
BenchmarkByIndexes5-4   	     297	   3642235 ns/op
```
```go
package internal

import (
	"fmt"
	"strconv"
	"strings"
	"testing"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/fields"
	"k8s.io/client-go/tools/cache"
)

func benchmarkByIndexes(sel fields.Selector, b *testing.B) {
	requires := sel.Requirements()
	indexFunc := func(val string) func(obj interface{}) ([]string, error) {
		return func(obj interface{}) ([]string, error) {
			pod := obj.(*corev1.Pod)
			if strings.Contains(pod.Labels["foo"], val) {
				return []string{KeyToNamespacedKey("", val)}, nil
			}
			return nil, nil
		}
	}

	index := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{
		FieldIndexName("idx1"): indexFunc("1"),
		FieldIndexName("idx2"): indexFunc("2"),
		FieldIndexName("idx3"): indexFunc("3"),
		FieldIndexName("idx4"): indexFunc("4"),
		FieldIndexName("idx5"): indexFunc("5"),
	})

	for i := 0; i < 15000; i++ {
		err := index.Add(&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
			Name: fmt.Sprintf("test-pod-%d", i), Labels: map[string]string{"foo": strconv.Itoa(i)}}})
		if err != nil {
			b.Fatal(err)
		}
	}

	for i := 0; i < b.N; i++ {
		objs, err := byIndexes(index, requires, "")
		if err != nil {
			b.Fatal(err)
		}
		if len(objs) == 0 {
			b.Fatal("expect to match objs, got 0")
		}
	}
}

func BenchmarkByIndexes1(b *testing.B) {
	sel := fields.ParseSelectorOrDie("idx1=1")
	benchmarkByIndexes(sel, b)
}

func BenchmarkByIndexes2(b *testing.B) {
	sel := fields.ParseSelectorOrDie("idx1=1,idx2=2")
	benchmarkByIndexes(sel, b)
}

func BenchmarkByIndexes3(b *testing.B) {
	sel := fields.ParseSelectorOrDie("idx1=1,idx2=2,idx3=3")
	benchmarkByIndexes(sel, b)
}

func BenchmarkByIndexes4(b *testing.B) {
	sel := fields.ParseSelectorOrDie("idx1=1,idx2=2,idx3=3,idx4=4")
	benchmarkByIndexes(sel, b)
}

func BenchmarkByIndexes5(b *testing.B) {
	sel := fields.ParseSelectorOrDie("idx1=1,idx2=2,idx3=3,idx4=4,idx5=5")
	benchmarkByIndexes(sel, b)
}
```

**Comment by tareksha** [comment]
@halfcrazy interesting so this indexing, and maybe k8s client-go indexes in general, does not really provide map-like lookup performance. it is a more of a convenience method.

**Pull request merged**